### PR TITLE
[Event Hubs Client - Track Two Preview 1] End-to-End Event Hub Client and Board Review

### PR DESCRIPTION
**Pull request by jsquire**
# Summary

The intent of these changes is to complete the end-to-end infrastructure of an `EventHubClient`, from the API surface through the compatibility shim for the track one implementation.  At this point, the client-level functionality such as querying Event Hub metadata is in-place and testable via the included live tests.  Also a core focus of these changes is incorporating the latest round of feedback from architectural board reviews, primarily aimed at tuning the public API surface.

# Goals

- Ensure that an `EventHubClient` creates a transport-aware client for containment to which it delegates operations.  This is currently used as a `TrackOneEventHubClient` and is intended to allow for concepts such as `AmqpEventHubClient` and `ReplayTestEventHubClient` to be used post-preview.

- Broker the compatibility between the track two API and the track one `EventHubClient` implementation, mapping types as needed.

- Allow for creation of an `EventHubClient` that is capable of interacting with the Event Hubs service for client-level operations, such as querying metadata.

- Address feedback  from the Azure Developer Platform Team architecture board.

- Review and revise the public surface area of the implementation, reducing scope as possible by internalizing constructs intended only to support project-level testing.

# Non-Goals

- Providing the implementation for the compatibility layer between the track two and track one `EventSender`; this will happen as part of future efforts.

- Providing the implementation for the compatibility layer between the track two and track one `EventReceiver`; this will happen as part of future efforts.

- Supporting validation or other authorization-related tasks for non-SAS credential types; this will happen as part of future efforts.

- Cover the identified set of end-to-end live tests needed to fully exercise the target scenarios; this will happen as part of future efforts.

- Inclusion of the full set of supporting assets, such as README, samples, and ARM templates; these will happen as part of future efforts.

- Enabling test parallelism; this will be a stretch goal, but unlikely to occur in the time frame for the first track two preview.

# Details  
  
### General
        
- Ongoing review and updates to member visibility to reduce public scope and limit protected members.

- Removal of some package references carried over from track one that have been determined to be unnecessary.

- Inclusion of the BCL interfaces preview package and supporting dependencies to allow for IAsyncDisposable use.

###   Architecture Board Review Feedback

- Hide `MaximumSizeInBytes` from the `EventBatchingOptions`, it is associated with a feature no longer in scope for the first preview.

- Rename of `ConnectionType` to `TransportType` to clarify semantics.

- Remove `RetrievalTimeUtc` from `EventData` to align members cross-language.

- Hide `SystemProperties` on `EventData` in order to present a more clear surface.  Properties from the system set have been hoisted to first class members of `EventData`.   At this time, it is believed that there are no additional metadata items sent by the service that we wish to expose to callers.

- Hide `Credential` and `ClientOptions` on `EventHubClient`; these are internal state tracking members used by the class and were not intended to be exposed to consumers.

- Rename `PartitionReceiver` to `EventReceiver` to better align terminology between senders and receivers.  The associated `CreatePartitionReceiver` member on `EventHubClient` was updated to `CreateReceiver` to reflect this change.

- Refactor to rename "BatchLabel" to "PartitionKey" in all contexts, along with documentation updates to try and add clarity to the meaning and semantics of it.

- Marked the `ReceiverOptions` member `DefaultMaximumReceiveWaitTime` as internal, allowing it to be hidden from consumers without causing churn and instability in the current flow of options.   _(This will be revisited post-preview to better understand if it is needed at all and, if so, whether there is benefit to exposing it to consumers.)_

###  Event Hub Client

- Refactoring to tighten access modifier scopes; testing approach has been adjusted to make greater use of custom mock objects and rely less on mocking via Moq to help limit the need for `protected` members exposed only to allow internal testing.

- Added support for IAsyncDisposable, delegating to `CloseAsync` to allow for more consise and representative usage patterns rather than requiring an explicit try/finally wrapper to ensure proper closure.

- Implemented construction of an inner transport client to use via containment as a delegate for performing the operations requested by consumers.

- Wired up delegation to the active transport client to service operations requested of the `EventHubClient`.

- Introduced Live tests for basic operations directly on the `EventHubClient`, absent creating senders and receivers.

###  Transport Client

 - Introduced as an abstraction to allow for a generalized Event Hub client interface specific to a protocol/transport to be used for delegation of operations.  It is intended that the public-facing `EventHubClient` create a dedicated transport client based on the active set of client options and make use if it via containment to avoid having protocol/transport concerns exposed.

- Created a transport client shim to serve as a bridge and adapter between the track two API and the track one implementation, allowing the preview to delegate operations to the track one types.

###  Authorization

- Added a default duration for shared key credentials in support of connection string use.

###  Test Infrastructure

- Added an abstraction for running Live tests in an `EventHubScope`, which is intended to allow for dynamic creation of an Event Hub with a given number of partitions, uniquely named.  This will allow Live tests to run strictly sandboxed, with a single test running against a give Hub.

    **NOTE:** _This has not been fully implemented yet; currently, the structure is in place to allow for creating Live tests in the form that will be needed to avoid a large and sweeping refactoring when the management functionality is added._

# Last Upstream Rebase

Monday, June 10, 2019  11:06am (EDT)

# Resources

- [.NET Event Hubs Client: Track Two Proposal (First Preview)](https://gist.github.com/jsquire/75b3a3090b6b4553aa8ceb798b6fc87d)
- [.NET Event Hubs Client: Deferred Features and Functionality](https://gist.github.com/jsquire/f88922faa0f457b503234c6f168e20c9)

# Related and Follow-Up Issues

- [[Epic] Release Event Hubs Client Library Track Two Preview 1](https://github.com/Azure/azure-sdk-for-net/issues/6030) (#6030)

### CODE REVIEW

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventSender.cs [hunk 1]
```diff
@@ -67,9 +67,9 @@ public class EventSender
         ///   caller to ensure that any needed cloning of options is performed.
         /// </remarks>
         ///
-        protected internal EventSender(ConnectionType connectionType,
-                                       string eventHubPath,
-                                       SenderOptions senderOptions)
+        internal EventSender(TransportType connectionType,
+                             string eventHubPath,
+                             SenderOptions senderOptions)
         {
```
Comment (chrisjbarr) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventSender.cs]: Rename `connectionType` parameter to `transportType`

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventSender.cs [hunk 2]
```diff
@@ -138,6 +138,15 @@ protected EventSender()
         ///
         public virtual void Close(CancellationToken cancellationToken = default) => CloseAsync(cancellationToken).GetAwaiter().GetResult();
 
+        /// <summary>
+        ///   Performs the tassk needed to clean up resources used by the <see cref="EventHubClient" />,
+        ///   including ensuring that the client itself has been cloased.
+        /// </summary>
+        ///
```
Comment (chrisjbarr) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventSender.cs]: typo: `tassk`

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Resources.resx [hunk 3]
```diff
@@ -171,4 +171,7 @@
   <data name="ResourceMustMatchSharedAccessSignature" xml:space="preserve">
     <value>The reqested resource, '{0}', does not match the resource of the shared access signature, '{1}'.   A token cannot be issued.</value>
   </data>
+  <data name="InvalidConnectionType" xml:space="preserve">
```
Comment (chrisjbarr) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Resources.resx]: Should this resource now be `InvalidTransportType`? along with an updated message `The requested transport type, '{0}' is not supported.`?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/TransportType.cs [hunk 4]
```diff
@@ -4,11 +4,11 @@
 namespace Azure.Messaging.EventHubs
 {
     /// <summary>
-    ///   Specifies the type of connection what will be used for communicating with
+    ///   Specifies the type of protcol and transport that will be used for communicating with
```
Comment (chrisjbarr) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/TransportType.cs]: typeo: `protcol`

**Review by chrisjbarr — CHANGES_REQUESTED** [review]

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubClient.cs [hunk 5]
```diff
@@ -271,6 +293,15 @@ public virtual EventSender CreateSender(SenderOptions senderOptions = default)
...
         public virtual void Close(CancellationToken cancellationToken = default) => CloseAsync(cancellationToken).GetAwaiter().GetResult();
 
+        /// <summary>
+        ///   Performs the task needed to clean up resources used by the <see cref="EventHubClient" />,
+        ///   including ensuring that the client itself has been closed.
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        public virtual async ValueTask DisposeAsync() => await CloseAsync();
```
Comment (AlexGhiondea) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubClient.cs]: Should this have a `ConfigureAwait(false)` ?

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubClient.cs]: Yes.... yes, it should.

**Review by AlexGhiondea — APPROVED** [review]

**Pull request merged**